Kernel/VirtIO/VirtIOQueue.cpp

```diff
@@ -130,4 +119,74 @@ bool VirtIOQueue::can_write() const
     return m_free_buffers > 0;
 }
 
+SpinLock<u8>& VirtIOQueue::lock()
+{
+    return m_lock;
+}
+
+bool VirtIOQueueChain::add_buffer_to_chain(FlatPtr buffer_start, size_t buffer_length, BufferType buffer_type) {
+    VERIFY(m_queue.lock().is_locked());
+
+    // Check that queue has space
+    if (!m_queue.can_write()) return false;
+
+    // Take a free slot from the queue
+    auto descriptor_index = m_queue.m_free_head;
+    m_queue.m_free_head = m_queue.m_descriptors[descriptor_index].next;
+    --m_queue.m_free_buffers;
+
+    // Ensure that no readable pages have been inserted after a writable one, as required by the VirtIO spec
+    VERIFY(buffer_type != BufferType::DeviceWritable || !m_chain_has_writable_pages);
+    m_chain_has_writable_pages |= buffer_type == BufferType::DeviceWritable;
+
+    // Link from previous element in VirtIOQueueChain
+    if (m_end_of_chain_index.has_value()) {
+        m_queue.m_descriptors[m_end_of_chain_index.value()].flags |= VIRTQ_DESC_F_NEXT;
+        m_queue.m_descriptors[m_end_of_chain_index.value()].next = descriptor_index;
+    }
+
+    // Set start of chain if it hasn't been set
+    if (!m_start_of_chain_index.has_value())
+        m_start_of_chain_index = descriptor_index;
+
+    // Update end of chain
+    m_end_of_chain_index = descriptor_index;
+    ++m_chain_length;
+
+    // Insert buffer
+    m_queue.m_descriptors[descriptor_index].address = static_cast<u64>(buffer_start);
+    VERIFY(buffer_length <= NumericLimits<size_t>::max());
+    m_queue.m_descriptors[descriptor_index].length = static_cast<u32>(buffer_length);
+
+    return true;
+}
+
+void VirtIOQueueChain::commit() {
+    VERIFY(m_queue.lock().is_locked());
+    VERIFY(m_start_of_chain_index.has_value());
+
+    auto next_index = m_queue.m_driver->index % m_queue.m_queue_size;
+    m_queue.m_driver->rings[next_index] = m_start_of_chain_index.value();
+    full_memory_barrier();
+    ++(m_queue.m_driver->index);
```

Comment:
is there any reason for removing the shadow index? accessing the device's index causes hardware interrupts to the device, which will be wasted since you only update the ring after the first access

Reply:
Oh, that's why the shadow index was there.

Reply:
I'll add that back in.